elastic/kibana
# Revert "Revert "[Upgrade Assistant] Refactor telemetry (#112177)" (#113665)"

This reverts commit c385d498874d4ca34f454e3cb9ad9e4c0e3219ae (https://github.com/elastic/kibana/pull/113665) and fixes the bug described in that PR. Reimplements https://github.com/elastic/kibana/pull/112177.

## How to review

The first commit is a mirror of the changes introduced in https://github.com/elastic/kibana/pull/112177, except for a couple minor divergences due to merge conflicts. You could probably pull up the diff from this PR side-by-side with that diff and just compare red/green patterns to ensure the code looks good.

The second and third commits are very small and should be straightforward to review.

## Steps to test

Because this change is a reimplementation of https://github.com/elastic/kibana/pull/112177, the steps to test are the same as that one, with a couple additions:

* Create a 6.8 sample data set that contains `ui_open` attributes in the UA telemetry saved object by visiting Upgrade Assistant and navigating to the different tabs. Verify the saved object has this attribute by going to Advanced Settings, filtering down to "Telemetry" and clicking the "See an example of what we collect". A flyout will open. Search the JSON in the flyout for `upgrade-assistant-telemetry` and verify that some of the `ui_open` attributes have non-zero values.
* Start up ES using this data. Assuming your data is located at`kibana/.es/6.8-data`, this would be by running `yarn es snapshot -E path.data=../6.8-data`.
* Verify telemetry works by performing the testing steps described in https://github.com/elastic/kibana/pull/112177.
* As you do so, verify that Kibana starts normally without crashing. This indicates that the saved object migration I implemented to fix the bug reported in https://github.com/elastic/kibana/pull/113665 is working as expected.
